pkg/container/ring_buffer.go

```diff
@@ -0,0 +1,121 @@
+package container
+
+import (
+	"sort"
+)
+
+// RingBuffer is a generic ring buffer implementation that contains
+// sequential data (i.e. such as time ordered data).
+// RingBuffer is implemented using slices. From testing, this should
+// be fast than linked-list implementations, and also allows for efficient
+// indexing of ordered data.
+type RingBuffer[T any] struct {
+	buffer  []T
+	index   int // index of ring buffer head.
+	maxSize int
+}
+
+func NewRingBuffer[T any](bufferSize int) *RingBuffer[T] {
+	return &RingBuffer[T]{
+		buffer:  make([]T, 0, bufferSize),
+		maxSize: bufferSize,
+	}
+}
+
+func (eb *RingBuffer[T]) isFull() bool {
+	return len(eb.buffer) >= eb.maxSize
+}
+
+func (eb *RingBuffer[T]) incr() {
+	eb.index = (eb.index + 1) % len(eb.buffer)
+}
+
+// Add adds an element to the buffer.
+func (eb *RingBuffer[T]) Add(e T) {
+	if eb.maxSize == 0 {
+		return
+	}
+	if eb.isFull() {
+		eb.incr()
+		eb.buffer[eb.index%len(eb.buffer)] = e
+		return
+	}
+	eb.buffer = append(eb.buffer, e)
+	eb.incr()
+}
+
+func (eb *RingBuffer[T]) dumpWithCallback(callback func(v T)) {
+	for i := 0; i < len(eb.buffer); i++ {
+		callback(eb.at(i))
+	}
+}
+
+func (eb *RingBuffer[T]) at(i int) T {
+	v := eb.buffer[(eb.index+1+i)%len(eb.buffer)]
+	return v
+}
+
+func (eb *RingBuffer[T]) firstValidIndex(isValid func(T) bool) int {
+	return sort.Search(len(eb.buffer), func(i int) bool {
+		return isValid(eb.at(i))
+	})
+}
+
+// IterateValid calls the callback on each element of the buffer, starting with
+// the first element in the buffer that satisfies "isValid".
+func (eb *RingBuffer[T]) IterateValid(isValid func(T) bool, callback func(T)) {
+	if !eb.isFull() {
+		for _, e := range eb.buffer {
+			callback(e)
+		}
+		return
+	}
+	startIndex := eb.firstValidIndex(isValid)
+	l := len(eb.buffer) - startIndex
+	for i := 0; i < l; i++ {
+		index := (eb.index + 1 + startIndex + i) % len(eb.buffer)
+		callback(eb.buffer[index])
+	}
+}
+
+// maps index in [0:len(buffer)) to the actual index in buffer.
+func (eb *RingBuffer[T]) mapIndex(index int) int {
+	tail := eb.index + 1
+	return (tail + index) % len(eb.buffer)
+}
+
+// GC clears out invalidated elements in the buffer.
+// This may require copying the entire buffer.
+// It is assumed that if buffer[i] is invalid then every entry [0...i-1] is also not valid.
+func (eb *RingBuffer[T]) GC(isValid func(T) bool) {
```

Comment:
bikeshed: GC is an overloaded term. maybe Prune or Compact?

Reply:
Agreed, I like compact ... or cull?